### ISSUE TITLE
Add support for -parallel=[true|false|NUM]

### ISSUE THREAD

**Issue by ladar** [hashicorp/packer#5615]
Low priority feature request... but it would be nice if the parallel command line option accepted a number, which represented the number of parallel builds. In other words:

```
1 == false
X == 1 < x < all/true
# == true or the number of boxes configured
```

My build bots can handle 2-8 parallel builds (depending on hardware, builder, etc), but if I enable parallel builds, and it tries to build all 20 at once, I run into problems with script timeouts, etc. It would be nice if there was some way to enable parallel builds, but limit that number. 

If this feature gets created, it might make sense to make the parallel option default to the number of CPUs on a system...

**Comment by grv87** [comment]
One build could use several CPUs.
So, it would be better to pass the number of CPUs Packer can use, and:
* Packer could ask VM engine (VirtualBox etc.) how much CPUs are used by VM being built
* (for e.g. Virtualbox) make `cpus` first-class configuration (without `vboxmanage`)
And use these numbers to decide, before starting another build, are there enough free CPUs left. If not then wait.
Also, this would require some *planner* that could decide e.g. to run builds that require maximum number of CPUs first.

By default use all available CPUs.

Also, negative number would mean leave this number of CPUs free.

**Comment by ladar** [comment]
@grv87 I like your idea, but it might be hard to implement support for what you described across all providers. Perhaps an initial schema that simply uses the number of physical cores (aka not the extra virtual/hyper-threaded cores), and by default, with only run that many builds at once?

What I found is that, in general, most of what packer does will run just fine with 25+ builds running, and the system at 100%. Unfortunately it's the `boot_command` piece that was prone to failure. When a system is first booted, the timing used by the `boot_command` and `boot_wait` parameters catch, and then simulate typing the commands needed by a particular OS installer to run unattended - is a very fickle phase. Tolerances vary, but there are a few distros which only provide a 5 to 10 second window. I'd say, off the top of my head, that of the 27 distros I'm currently building, half of them work work decently even if the system is under load.  Unfortunately it's the other half, the party poopers, which force me to use `-parallel=false`, and handle parallelization manually.

The other issue I found is that with some hypervisors, when a system is under heavy load, the key press events sent by the `boot_command` won't register insides a guest VM. I researched this, and found that it was probably because the `key down` and `key up` events are occurring before the OS scheduler gives the guest enough cycles to process the input. This was a show stopper for me because it breaks (or worse, corrupts) the build (aka OS installation) process. It was also hard, because again some hypervisors/packer builders worked better than others, and some guests, did better than others. 

I've had limited success elevating the priority of packer, and a guest process during the initial boot phase, but I don't think that tactic will work well in code (a normal user account doesn't have permission to elevate a process beyond the norm, and halting to ask for sudo/admin would defeat the purpose). 

I should also mention that I found the vagrant and/or compress post processors to be the biggest offenders. If a `vagrant` and/or `compress` post processor is running when a builder is performing the initial boot, the problems described above are far more likely to happen. I personally found a workaround to this issue, somewhat, by using the GOMAXPROCS environment variable. It allows me to control how CPUs a single build job will consume, but it's an inefficient strategy. (I can't adjust the value once packer starts, so if 20 builds are finished, the stragglers still only use 2 cores, instead of 12 to 24 that might be available.

I mentioned the GOMAXPROCS variable, because it might solve your problem. It will, at least, allow you to control how many CPUs packer consumes (per build). That keeps CPU hungry operations like `vagrant`, and `compress` from overwhelming the system. But sadly, it won't let you control how many independent builds can run.

**Comment by grv87** [comment]
@ladar, I see. My proposition is hard to implement but it has small utility.

I know about `boot_command` problem. I couldn't have made work it under Windows host and ended up building *kernel* images (just preinstalled OS without any extra configuration) under Debian or Ubuntu and then building *base* boxes from *kernel* under any host available.

I'm currently working on Gradle plugin for Packer.
Gradle is a general-purpose build system.
Plan is not to use Packer parallelism, but to have Gradle task for each single build and to allow Gradle to run them in parallel if it is appropriate. Each task could have its inputs and outputs, and other task dependencies. Gradle don't have ability to configure parallelism too. But at least it allows to manually configure that some tasks should run after some other tasks. So, it is a way to limit parallelism.
I see there is some overhead to run several packer processes for builds from single template but I think it's worth that.
My plugin will also have up-to-date detection (incremental building) which Packer doesn't have. I'm not sure how useful it will be if you rebuild images by schedule. But I found it very useful during development. If you change two configuration options in one builder, you don't want to rebuild everything.

**Comment by azr** [comment]
While this is a breaking change I think this is a good idea; I'd go for just a limit in the number of parallel builds; then a user ca fine grain this depending on their use case and environment.

This can be controller in this file:
https://github.com/hashicorp/packer/blob/e666b60d166db7b4c6185d8b9a5417c6622d68c7/command/build.go#L144-L152

Where the loop waits if we are debugging or running with parallel=false
https://github.com/hashicorp/packer/blob/e666b60d166db7b4c6185d8b9a5417c6622d68c7/command/build.go#L198-L201

One thing you could do is use channels with sizes:


```go
c := make(chan struct{}, parallelism)

go func(){
    c <- struct{}{} // there can only be only  'parallelism' values in c otherwise this blocks
    defer func() { <-c }()
    //...
}
```

**Comment by azr** [comment]
hello here, we have a candidate PR in #7501, and it's looking good to me, also I edited it a little to make it a non breaking change, so this can be merged soon.
Opinions welcome ! 🙂  personally I really like @/wandel's option to go for semaphores !
